lib/hyacinth/digital_object/type_def/dynamic_field_data.rb

```diff
@@ -0,0 +1,117 @@
+# frozen_string_literal: true
+
+module Hyacinth
+  module DigitalObject
+    module TypeDef
+      class DynamicFieldData < Hyacinth::DigitalObject::TypeDef::JsonSerializableCollection
+        TYPE_TO_FORM_TYPE = {
+          rights_metadata: ['item_rights', 'asset_rights'],
+          descriptive_metadata: ['descriptive']
+        }.freeze
+
+        def initialize(type)
+          super()
+          raise ArgumentError, "type '#{type}' is invalid" unless TYPE_TO_FORM_TYPE.keys.map(&:to_s).include?(type.to_s)
+
+          @type = type.to_sym
+        end
+
+        def from_serialized_form_impl(json_var)
+          return nil if json_var.nil?
+          raise ArgumentError, "Expected hash, but got: #{json_var.class}" unless json_var.is_a?(Hash)
+
+          terms_map = extract_terms(field_map, json_var)
+
+          search_query = terms_map.map { |vocab, terms|
+            uris = terms.map { |t| "\"#{t['uri']}\"" }.compact.join(' OR ')
+            "(vocabulary:\"#{vocab}\" AND uri:(#{uris}))"
+          }.join(' OR ')
+
+          search_results = Hyacinth::Config.term_search_adapter.search(post: true) do |solr_params|
+            solr_params.q(search_query, escape: false)
+          end
+
+          # Retrieve all custom fields for vocabularies so we know which fields should be add to the hash.
+          # This ensure that there's always a value for the all the custom fields defined and does not add
+          # any custom fields that have been removed from the Vocabulary since indexing.
```

Comment:
Does this comment mean that we're expecting that `Hyacinth::Config.term_search_adapter.search` could theoretically return search results that include no-longer-existing custom fields or are missing existing custom fields?  If so, that sounds like a job that the search adapter should be doing rather than handling it only in the typedef, so we benefit from this inconsistency-handling app-wide.  (Unless I'm misunderstanding?)

Reply:
Yes, if a custom_field is added/removed from a vocabulary all the terms for that vocabulary aren't updated to reflect that change. I wouldn't expect it often, but it could happen. It seems like the ideal solution would be to reindex the terms again after a vocabularies custom field change. We could update the `custom_fields` hash in `Hyacinth::Config.term_search_adapter.search`, but we would additionally also have to update the custom field's indexed fields.

Reply:
Rather than reindexing the terms after a custom field change, what if we disallow deletion until the used values are cleared out?

The reason I like that approach is that it resembles a few other Hyacinth cases (that aren't yet implemented) for batch-scanning records when certain global changes occur:
- If someone attempts to delete a dynamic field, scan through DigitalObject solr docs to determine whether that field is in use and reject deletion if it is.
- If someone attempts to delete a project, do the same thing to make sure the project isn't in use.
- If someone wants to delete a controlled term from a vocabulary, do the same thing, making sure that term isn't referenced by any records.

And now we would have this other case for making sure that if someone attempts to delete a controlled term's custom field, we first determine if any terms are using that custom field.

Are custom fields with blank values absent from the solr document?  If so, then I think what I'm proposing would work and we wouldn't have to do any kind of additional reindexing of terms *after* deletion.  So to recap:

To clear a vocabulary's custom field that has NO occurrences in any terms:
1) delete operation first checks if it's present in any terms, which is isn't
2) custom field is deleted from vocabulary.  done.

To clear a vocabulary's custom field that HAS occurrences in some terms:
1) delete operation first checks if it's present in any terms, which it is. deletion rejected.
2) a Hyacinth user is responsible for clearing the field from the terms that use it (knowable via solr query) in whatever way we see fit (i.e. actually delete it OR migrate the value to a new custom field if applicable)
3) the updating of a term cascades and updates all of the digital objects that reference it (via reindexing of those digital objects) -- so this process depends on the change-term-and-reindex-digital-objects functionality we don't have yet, but that we need.
4) we retry the vocabulary custom field deletion again, and it succeeds because no terms reference the custom field anymore.

What do you think?  If this sounds reasonable, I can ticket it.

And then it would potentially still make sense to move the logic I mentioned in my earlier comment to the term search adapter.

Reply:
One thing to consider: custom fields with blank values aren't always absent from the database record and the solr record if they are empty (at least not right now). Because we store custom fields in a hash if a `blank?` value has been set the key of that field is most likely present in the custom fields hash and has therefore been indexed. We index the custom fields json into a solr field. In order to assure that there are no blank values in the custom fields we would have to apply logic to that effect. 

I think the rests works but its very theoretical now. What are we going to implement in the short term? Remove the ability to delete custom fields?

Reply:
For the short-term having no UI ability to delete custom fields works.  I think we may want to clear out empty custom field values in the near future, and then we'll work through the other aspects of this problem later on.

Reply:
Okay, sounds good. But the logic here would stay the same, correct? or are you suggesting changes to this PR?

Reply:
I'm good with your suggestion to make this future work, so no changes to this PR.

Reply:
👍